geoportal/c2cgeoportal_geoportal/views/entry.py

```diff
@@ -905,6 +933,41 @@ def _wms_get_features_type(self, ogc_server_id, wfs_url):
             ))
             return None, errors
 
+        return response, errors
+
+    def get_url_internal_wfs(self, ogc_server, errors):
+        # required to do every time to validate the url.
+        if ogc_server.auth != main.OGCSERVER_AUTH_NOAUTH:
+            url = self.request.route_url("mapserverproxy", _query={"ogcserver": ogc_server.name})
+            url_wfs = url
+            url_internal_wfs = get_url2(
+                "The OGC server (WFS) '{}'".format(ogc_server.name),
+                ogc_server.url_wfs or ogc_server.url, self.request, errors=errors
+            )
+        else:
+            url = get_url2(
+                "The OGC server '{}'".format(ogc_server.name),
+                ogc_server.url, self.request, errors=errors
+            )
+            url_wfs = get_url2(
+                "The OGC server (WFS) '{}'".format(ogc_server.name),
+                ogc_server.url_wfs, self.request, errors=errors
+            ) if ogc_server.url_wfs is not None else url
+            url_internal_wfs = url_wfs
+        return url_internal_wfs, url, url_wfs
+
+    def preload(self, errors):
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+        tasks = set()
+        for ogc_server in models.DBSession.query(main.OGCServer).all():
+            url_internal_wfs, _, _ = self.get_url_internal_wfs(ogc_server, errors)
+            tasks.add(self._async_wms_get_features_type(ogc_server.id, url_internal_wfs))
+            tasks.add(self._async_wms_getcap(ogc_server))
+
+        loop.run_until_complete(asyncio.gather(*tasks))
```

Comment:
Why not asyncio.run, python doc says:
Application developers should typically use the high-level asyncio functions, such as asyncio.run(), and should rarely need to reference the loop object or call its methods.

Reply:
It's the first time I use asyncio than it's possible that I do some little mistake, but here if I use `asyncio.run` in place of `loop.run_until_complete` I get:
`ValueError: a coroutine was expected, got <_GatheringFuture pending>`
...

Reply:
If I have a look of the doc:
https://docs.python.org/3/library/asyncio-task.html#asyncio.gather
To use run I should write a new method, then I prefer like that :-)

Reply:
You should not do give the gather to asyncio.run but an async function. No need for a new method: 
```
async def preload(self):
    ...
    await asyncio.gather(*tasks)

def _themes(self):
    asyncio.run(self.preload())
```

Reply:
Yes but it moves many things to the theme_ method :-/